### PR TITLE
feat: custom DJ Role for parrot commands

### PR DESCRIPTION
**Pull request by joao-conde**
| - | - |
| --- | --- |
| Issue | Closes https://github.com/aquelemiguel/parrot/issues/47 |
| Dependencies | `"Manage Roles"` permission is now required and Wiki should be updated after this merge. |
| Decisions | - Refactor code to make handlers smaller and call what they need.<br>- Set permissions for all guilds for all commands of the DJ Role.<br>- The DJ Role name can then be changed by the discord admins if they want, but I picked a default ("Parrot DJ").<br>- Parrot ensures the role exists by either using an already existing one or creating a new one.<br>- now playing, queue, play, summon and version are unrestrained |
| Animated GIF | dj-role-commands |

### CODE REVIEW

File: src/handlers/serenity.rs [hunk 1]
```diff
@@ -17,45 +22,102 @@ use crate::commands::{
     remove::*, repeat::*, resume::*, seek::*, shuffle::*, skip::*, stop::*, summon::*, version::*,
 };
 
+const ROLE: &str = "Parrot DJ";
```
Comment (afonsojramos) [inline src/handlers/serenity.rs:25]: nitpick: Is it possible to get the bot's name here instead of simply "Parrot DJ"?

**Review by afonsojramos — CHANGES_REQUESTED** [review]

File: src/handlers/serenity.rs [hunk 2]
```diff
@@ -17,45 +22,102 @@ use crate::commands::{
...
                     command
                         .name("pause")
                         .description("Pauses the current track")
+                        .default_permission(false)
                 })
                 .create_application_command(|command| {
                     command
                         .name("play")
                         .description("Adds a track to the queue")
+                        .default_permission(false)
```
[the remark this answers is not in the document]
Comment (afonsojramos) [reply src/handlers/serenity.rs:120]: Summoning @aquelemiguel @rafaeldamasceno @danrpinho for opinions.

File: src/handlers/serenity.rs (hunk 1 above)
Comment (joao-conde) [reply src/handlers/serenity.rs:25]: The problem is that the bot's name is already its own role and can't be assigned to users. I also don't like this default but have no better idea. Thankfully it is also changeable by the Server Admin.

image

Comment (afonsojramos) [reply src/handlers/serenity.rs:25]: I meant something like "${botName}'s DJ".

File: src/handlers/serenity.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (afonsojramos) [reply src/handlers/serenity.rs:120]: I mean, the development is done now, so it's a matter of simply deciding what we want, and it is a small change 🤷

File: src/handlers/serenity.rs (hunk 1 above)
Comment (joao-conde) [reply src/handlers/serenity.rs:25]: Seems like a good idea I will look into it

File: src/handlers/serenity.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (joao-conde) [reply src/handlers/serenity.rs:118]: Resolving for now. Decide in the issue and I will check the conclusions.

**Review by afonsojramos — APPROVED** [review]

**Comment by afonsojramos** [comment]
image

Love the fact that this gets greyed out!

File: src/handlers/serenity.rs [hunk 3]
```diff
@@ -116,63 +181,78 @@ impl EventHandler for SerenityHandler {
...
     }
 
-    async fn voice_state_update(
-        &self,
-        ctx: Context,
-        guild: Option<GuildId>,
-        _old: Option<VoiceState>,
-        new: VoiceState,
-    ) {
+    async fn run_command(&self, ctx: &Context, command: &mut ApplicationCommandInteraction) {
```
Comment (aquelemiguel) [inline src/handlers/serenity.rs:231]: This isn't really a `run_command` but a more like a `register_command_callback` I'd say.

**Review by aquelemiguel — APPROVED** [review]
Small suggestion, but overall a high quality PR. Good stuff.

File: src/handlers/serenity.rs (hunk 3 above)
Comment (joao-conde) [reply src/handlers/serenity.rs:231]: No, this actually runs the command.

https://github.com/aquelemiguel/parrot/blob/jc/47-dj-permission/src/handlers/serenity.rs#L233

This actually runs the `autopause` command

**Review by afonsojramos — APPROVED** [review]

**Pull request merged**